src/Individuals/Abiotic/particle_interaction.jl

```diff
@@ -12,64 +12,114 @@
-end
-function calc_merge_matrix!(intac, abiotic, plank, rnd, grid, abio_p, arch::Architecture)
-    kernel! = calc_merge_matrix_kernel!(device(arch), (16,16), (size(intac)))
-    kernel!(intac, abiotic, plank, rnd, grid, abio_p)
+function reset_interaction!(ids, arch::Architecture)
+    kernel! = reset_interaction_kernel!(device(arch), (16, 16), size(ids))
+    kernel!(ids)
     return nothing
 end
 
-##### clean the merge matrix - one vesicle might be close enough
-##### to two or more phytoplankton cells. 
-##### Only the first phytoplankton cell can uptake this vesicle.
-##### First, find all indices of 1.0f0 in intac, return an AbstracArray of
-##### Cartesian Index, e.g., CartesianIndex(phyto_ind, abiotic_ind)
-##### Then, assign phyto_ind to abiotic.merge[abiotic_ind]
-@kernel function assign_merge_kernel!(abiotic, inds)
+##### 2. Calculate Interaction Topology 
+##### Revised: Uses random slot assignment to avoid race conditions.
+@kernel function calc_interaction_kernel!(top_ids, plank, abiotic, rnd, grid, abio_p, K)
     i = @index(Global)
-    abiotic.merg[inds[i][2]] = inds[i][1]
+    
+    # Check if phytoplankton is active
+    if plank.ac[i] == 1.0f0 
+        
+        my_rnd = rnd.x[i]
+        
+        for j in 1:length(abiotic.ac)
+            if abiotic.ac[j] == 1.0f0
+                dist = calc_distance(plank.xi[i], plank.yi[i], plank.zi[i], plank.x[i], plank.y[i], plank.z[i],
+                                     abiotic.xi[j], abiotic.yi[j], abiotic.zi[j], abiotic.x[j], abiotic.y[j], abiotic.z[j],
+                                     grid)
+                
+                # Probabilistic check (Distance < Sensing Radius * Random)
+                if dist < (abio_p.Rd * my_rnd)
+                    
+                    slot = (i % K) + 1
+                    
+                    # Simple overwrite: The last thread to write wins.
+                    # Since all threads here passed the probability check, any of them is a valid candidate.
+                    top_ids[j, slot] = i
+                end
+            end
+        end
+        
+    end 
 end
-function assign_merge!(abiotic, inds, arch::Architecture)
-    kernel! = assign_merge_kernel!(device(arch), 256, (size(inds,1)))
-    kernel!(abiotic, inds)
+
+function calc_interaction!(top_ids, plank, abiotic, rnd, grid, abio_p, K, arch::Architecture)
+    kernel! = calc_interaction_kernel!(device(arch), 256, (size(plank.ac, 1)))
+    # Removed top_dists from arguments
+    kernel!(top_ids, plank, abiotic, rnd, grid, abio_p, K)
     return nothing
 end
 
-##### merge particles
-@kernel function merge_particle_kernel!(abiotic, plank)
-    i = @index(Global)
-    if abiotic.merg[i] ≠ 0
-        @inbounds KernelAbstractions.@atomic plank.ptc[abiotic.merg[i]] += 1.0f0
+##### 3. Consume Particle 
+##### Handles actual uptake with capacity limit. Unchanged logic.
+@kernel function consume_particle_kernel!(top_ids, plank_ptc, abio_ac, max_uptake, K)
```

Comment:
directly use `plank` in function argument, and use`plank.ptc` in the function

Reply:
same for `abiotic.ac`

Reply:
Don't know if there's a performance difference. But if so, we'll need to rewrite the code...